# User Engagement Analysis
- User Engagement Analysis focuses on tracking user activities on database sessions to appreciate user engagement for overall applications and per application. The engagement metrics to be used are:
    - Sessions frequency
    - Duration of the session
    - Session total traffic (download and upload in bytes)


In [1]:
# Import necessary libraries and modules
import os
import sys
import warnings
import pandas as pd

# Add parent directory to path to import local modules
sys.path.insert(0, os.path.dirname(os.getcwd()))
from scripts.user_engag_analysis import UserEngagementAnalysis


# Suppress FutureWarnings
warnings.simplefilter(action='ignore', category=FutureWarning)

## Load and Display Cleaned Data

In [2]:
# Load the data
df = pd.read_csv('../data/clean_proccessed_data.csv')

# Display the first few rows of the DataFrame
df.head()

Bearer Id            Start              End   Dur. (s)  MSISDN/Number  \
0  1.311448e+19   4/4/2019 12:01  4/25/2019 14:35  1823652.0   3.366496e+10   
1  1.311448e+19   4/9/2019 13:04   4/25/2019 8:15  1365104.0   3.368185e+10   
2  1.311448e+19   4/9/2019 17:42  4/25/2019 11:58  1361762.0   3.376063e+10   
3  1.311448e+19   4/10/2019 0:31   4/25/2019 7:36  1321509.0   3.375034e+10   
4  1.311448e+19  4/12/2019 20:10  4/25/2019 10:40  1089009.0   3.369980e+10   

           IMSI          IMEI  Avg Bearer TP DL (kbps)  \
0  2.082014e+14  3.552121e+13                     23.0   
1  2.082019e+14  3.579401e+13                     16.0   
2  2.082003e+14  3.528151e+13                      6.0   
3  2.082014e+14  3.535661e+13                     44.0   
4  2.082014e+14  3.540701e+13                      6.0   

   Avg Bearer TP UL (kbps)  Avg RTT DL (ms)  ...  Other DL (Bytes)  \
0                     44.0        42.000000  ...       171744450.0   
1                     26.0        65.000000  ...       526904238.0   
2                      9.0       109.795706  ...       410692588.0   
3                     44.0       109.795706  ...       749039933.0   
4                      9.0       109.795706  ...       550709500.0   

   Other UL (Bytes)  Total DL (Bytes)  Total UL (Bytes)  Handset Manufacturer  \
0         8814393.0       308879636.0        36749741.0               Samsung   
1        15055145.0       653384965.0        53800391.0               Samsung   
2         4215763.0       279807335.0        27883638.0               Samsung   
3        12797283.0       846028530.0        43324218.0             undefined   
4        13910322.0       569138589.0        38542814.0               Samsung   

                  Handset Type  DL TP < 50 Kbps (%)  \
0   Samsung Galaxy A5 Sm-A520F                100.0   
1  Samsung Galaxy J5 (Sm-J530)                100.0   
2     Samsung Galaxy A8 (2018)                100.0   
3                    undefined                100.0   
4             Samsung Sm-G390F                100.0   

   50 Kbps < DL TP < 250 Kbps (%)  UL TP < 10 Kbps (%)  \
0                             0.0                100.0   
1                             0.0                100.0   
2                             0.0                100.0   
3                             0.0                100.0   
4                             0.0                100.0   

   10 Kbps < UL TP < 50 Kbps (%)  
0                            0.0  
1                            0.0  
2                            0.0  
3                            0.0  
4                            0.0  

[5 rows x 35 columns]

### User Engagement Analysis Initialization

In [3]:
User_Engag = UserEngagementAnalysis(df)

In [4]:
aggregated_data= User_Engag.aggregate_metrics_per_customer()
print(aggregated_data.head())

   MSISDN/Number  session_frequency  session_duration  total_traffic
0   3.360100e+10                  1          116720.0   8.786906e+08
1   3.360100e+10                  1          181230.0   1.568596e+08
2   3.360100e+10                  1          134969.0   5.959665e+08
3   3.360101e+10                  1           49878.0   4.223207e+08
4   3.360101e+10                  4           37104.0   1.457411e+09


### Aggregated Metrics per Customer

In [7]:
top_10_frequency, top_10_duration, top_10_traffic = User_Engag.top_10_customers_per_metric(aggregated_data)

In [6]:
top_10_frequency

('top_10_frequency',
         MSISDN/Number  session_frequency  session_duration  total_traffic
 106853   4.188282e+10            1136356      7.255100e+07   5.317447e+11
 13526    3.362632e+10                324      8.791927e+06   7.971167e+09
 6437     3.361489e+10                289      9.966898e+06   8.846226e+09
 13180    3.362578e+10                289      1.855375e+07   8.499621e+09
 37052    3.365973e+10                256      4.035428e+06   7.705863e+09
 76363    3.367588e+10                225      4.865947e+06   7.891111e+09
 92923    3.376054e+10                225      9.279434e+06   8.514774e+09
 65118    3.366716e+10                169      8.744914e+06   5.618394e+09
 666      3.360313e+10                144      6.287761e+06   4.976195e+09
 1279     3.360452e+10                144      5.207990e+06   5.487855e+09)

In [9]:
top_10_duration

('top_10_duration',
         MSISDN/Number  session_frequency  session_duration  total_traffic
 106853   4.188282e+10            1136356      7.255100e+07   5.317447e+11
 13180    3.362578e+10                289      1.855375e+07   8.499621e+09
 6437     3.361489e+10                289      9.966898e+06   8.846226e+09
 92923    3.376054e+10                225      9.279434e+06   8.514774e+09
 13526    3.362632e+10                324      8.791927e+06   7.971167e+09
 65118    3.366716e+10                169      8.744914e+06   5.618394e+09
 50281    3.366284e+10                 81      6.614270e+06   4.000098e+09
 57160    3.366469e+10                 81      6.288730e+06   5.682007e+09
 666      3.360313e+10                144      6.287761e+06   4.976195e+09
 66029    3.366746e+10                 36      5.649882e+06   2.600522e+09)

In [10]:
top_10_traffic

('top_10_traffic',
         MSISDN/Number  session_frequency  session_duration  total_traffic
 106853   4.188282e+10            1136356      7.255100e+07   5.317447e+11
 6437     3.361489e+10                289      9.966898e+06   8.846226e+09
 92923    3.376054e+10                225      9.279434e+06   8.514774e+09
 13180    3.362578e+10                289      1.855375e+07   8.499621e+09
 13526    3.362632e+10                324      8.791927e+06   7.971167e+09
 76363    3.367588e+10                225      4.865947e+06   7.891111e+09
 37052    3.365973e+10                256      4.035428e+06   7.705863e+09
 63028    3.366646e+10                121      4.536757e+06   7.308501e+09
 92577    3.376041e+10                144      5.321667e+06   7.132371e+09
 57241    3.366471e+10                121      2.927785e+06   6.872018e+09)

### Top 10 Customers by Session Frequency

Key insights include:

- **High Session Frequency**: The customer with MSISDN/Number `4.188282e+10` has an exceptionally high session frequency of 1,136,356 sessions, indicating very frequent usage.
- **Session Duration**: Despite the high session frequency, the session duration varies significantly among the top customers. For example, the top customer has a total session duration of `7.255100e+07` seconds, while others have much lower durations.
- **Total Traffic**: The total traffic (download and upload in bytes) also varies, with the top customer generating `5.317447e+11` bytes of traffic, which is significantly higher than others.

This analysis helps identify the most engaged users in terms of session frequency, which can be useful for targeted marketing and improving user experience.

### Top 10 Customers by Session Duration

Key insights include:

- **Longest Session Duration**: The customer with MSISDN/Number `4.188282e+10` has an exceptionally long session duration of `7.255100e+07` seconds, indicating very high engagement.
- **Session Frequency**: Despite the long session duration, the session frequency varies among the top customers. For example, the top customer has a session frequency of 1,136,356 sessions, while others have much lower frequencies.
- **Total Traffic**: The total traffic (download and upload in bytes) also varies, with the top customer generating `5.317447e+11` bytes of traffic, which is significantly higher than others.

This analysis helps identify the most engaged users in terms of session duration, which can be useful for understanding user behavior and improving user experience.

### Top 10 Customers by Total Traffic

Key insights include:

- **Highest Total Traffic**: The customer with MSISDN/Number `4.188282e+10` has an exceptionally high total traffic of `5.317447e+11` bytes, indicating very high data usage.
- **Session Frequency**: The session frequency varies among the top customers. For example, the top customer has a session frequency of 1,136,356 sessions, while others have much lower frequencies.
- **Session Duration**: The session duration also varies significantly among the top customers. For example, the top customer has a session duration of `7.255100e+07` seconds, while others have much lower durations.

This analysis helps identify the most data-intensive users, which can be useful for network optimization and targeted marketing strategies.